examples/cis2-wccd-upgrade/src/lib.rs

```diff
@@ -0,0 +1,2030 @@
+    operators: StateSet<Address, S>,
+}
+
+/// The contract state.
+#[derive(Serial, DeserialWithState)]
+#[concordium(state_parameter = "S")]
+struct State<S> {
+    /// Address of the w_ccd proxy contract.
+    proxy_address:          Option<ContractAddress>,
+    /// Address of the w_ccd implementation contract.
+    implementation_address: Option<ContractAddress>,
+    /// FALSE if proxy and implementation addresses have not been set yet.
+    /// TRUE if proxy and implementation addresses have been set.
+    initialized:            bool,
+    /// The state of the one token.
+    token:                  StateMap<Address, AddressState<S>, S>,
+    /// Timestamp when contract is unpaused.
+    unpause_time:           Timestamp,
+}
+
+/// The implementation contract state.
+#[derive(Serial, DeserialWithState)]
+#[concordium(state_parameter = "S")]
+struct StateImplementation<S> {
+    /// The admin address can pause/unpause the contract.
+    admin:         Address,
+    /// Address of the w_ccd proxy contract.
+    proxy_address: Option<ContractAddress>,
+    /// Address of the w_ccd state contract.
+    state_address: Option<ContractAddress>,
+    /// FALSE if proxy and state addresses have not been set yet.
+    /// TRUE if proxy and state addresses have been set.
+    initialized:   bool,
+
+    /// TODO: These state variables will be removed out of this
+    /// `StateImplementation` struct once the getter/setter functions are
+    /// set up between the State contract and the Implementation contract.
+    /// The state of the one token.
+    token:        StateMap<Address, AddressState<S>, S>,
+    /// Timestamp when contract is unpaused.
+    unpause_time: Timestamp,
+}
+
+/// The parameter type for the contract function `unwrap`.
+/// Takes an amount of tokens and unwrap the CCD and send it to a receiver.
+#[derive(Serialize, SchemaType)]
+struct UnwrapParams {
+    /// The amount of tokens to unwrap.
+    amount:   ContractTokenAmount,
+    /// The owner of the tokens.
+    owner:    Address,
+    /// The address to receive these unwrapped CCD.
+    receiver: Receiver,
+    /// Some additional bytes to include in the transfer.
+    data:     AdditionalData,
+}
+
+/// The parameter type for the contract function `wrap`.
+///
+/// The receiver for the wrapped CCD tokens.
+#[derive(Serialize, SchemaType)]
+struct WrapParams {
+    /// The address to receive these tokens.
+    /// If the receiver is the sender of the message wrapping the tokens, it
+    /// will not log a transfer.
+    to:   Receiver,
+    /// Some additional bytes to include in a transfer.
+    data: AdditionalData,
+}
+
+/// The parameter type for the state contract function `initialize`.
+#[derive(Serialize, SchemaType)]
+struct InitializeStateParams {
+    /// Address of the w_ccd proxy contract.
+    proxy_address:          Option<ContractAddress>,
+    /// Address of the w_ccd implementation contract.
+    implementation_address: Option<ContractAddress>,
```

Comment:
Are these suppose to be options?
Would it make sense to initialize with any of these as None?

Reply:
yes, they are initialized with `None` when the `init` function is called on the smart contract. We don't know the `implementation` or the `proxy` addresses during the call of the `init` function.

Reply:
It might make sense for `State` to have `Option<_>`, but I don't think it makes sense for the parameter to have the same. I would assume that when you invoke `contract_state_initialize` you _must_ provide both the proxy address and the implementation address.

So something like this:
```
#[derive(Serialize, SchemaType)]
struct InitializeStateParams {
    /// Address of the w_ccd proxy contract.
    proxy_address:          ContractAddress,
    /// Address of the w_ccd implementation contract.
    implementation_address: ContractAddress,
}
```

```
    let params: InitializeStateParams = ctx.parameter_cursor().get()?;
    host.state_mut().proxy_address = Some(params.proxy_address);
    host.state_mut().implementation_address = Some(params.implementation_address);
```

Reply:
And that would also be the case for `InitializeImplementationParams` 😊